File: sdk/src/main/java/com/cxense/cxensesdk/ApiCallback.java
package com.cxense.cxensesdk;

import androidx.annotation.NonNull;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Basic callback for API
 *
 * @param <T> success data type
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2017-06-05).
 */
class ApiCallback<T> implements Callback<T> {
    private final LoadCallback<T> callback;
    private final ApiErrorParser errorParser;

    ApiCallback(LoadCallback<T> callback, ApiErrorParser errorParser) {
        this.callback = callback;
        this.errorParser = errorParser;
    }

    @Override
    public void onResponse(@NonNull Call<T> call, @NonNull Response<T> response) {
        if (callback == null)
            return;

        if (response.isSuccessful()) {
            callback.onSuccess(response.body());
        } else callback.onError(errorParser.parseError(response));
    }

    @Override
    public void onFailure(@NonNull Call<T> call, @NonNull Throwable throwable) {
        if (callback != null)
            callback.onError(throwable);
    }
}

File: sdk/src/main/java/com/cxense/cxensesdk/ApiCallback.kt
package com.cxense.cxensesdk

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class ApiCallback<T>(
    private val callback: LoadCallback<T>,
    private val errorParser: ApiErrorParser
) : Callback<T> {
    override fun onFailure(call: Call<T>, t: Throwable) {
        callback.onError(t)
    }

    override fun onResponse(call: Call<T>, response: Response<T>) {
        val body = response.body()
        if (response.isSuccessful && body != null)
            callback.onSuccess(body)
        else {
            val e = errorParser.parseError(response) ?: BaseException("Response body is null")
            callback.onError(e)
        }
    }
}